virtualpdu/core.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2016 Internap
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+POWER_ON = 'POWER_ON'
+POWER_OFF = 'POWER_OFF'
+REBOOT = 'REBOOT'
+
+
+class Core(object):
+    def __init__(self, driver, mapping):
+        self.driver = driver
+        self.mapping = mapping
+        self.logger = logging.getLogger(__name__)
+
+    def _get_server_name(self, pdu_name,
+                         outlet_number):
+        return self.mapping[(pdu_name, outlet_number)]
+
+    def pdu_outlet_state_changed(self, name, outlet_number, state):
+        self.logger.warning(
+            "PDU '{name}', outlet '{outlet_number}' has new state: '{state}'".
+            format(name=name, outlet_number=outlet_number, state=state))
+        try:
+            server_name = self._get_server_name(name, outlet_number)
+
+            self.logger.warning("Found server '{0}'".format(server_name))
+        except KeyError:
+            return
+
+        if state == POWER_ON:
+            self.driver.power_on(server_name)
+        elif state == POWER_OFF:
+            self.driver.power_off(server_name)
+        elif state == REBOOT:
+            self.driver.power_off(server_name)
```

Comment:
future improvement, add LibvirtDriver:reboot and use [libvirt's reboot](https://libvirt.org/sources/java/javadoc/org/libvirt/Domain.html#reboot%28int%29)